### PR TITLE
Filter for container collection

### LINKED ISSUE #10: Create filter for RDF containers and collections
It would be nice to have a filter, which allows easy iteration over RDF containers and collections.

https://www.w3.org/TR/2014/REC-rdf11-mt-20140225/#rdf-containers
https://www.w3.org/TR/2014/REC-rdf11-mt-20140225/#rdf-collections

### LINKED ISSUE #66: Collections can't be rendered using a property to reference
If I try to reference a collection using a property I get an exception.
I'm on the feature branch from the pullrequest #62 .

My data:

```
<http://example.org/Sebastian>
    a <http://xmlns.com/foaf/0.1/Agent>, <http://xmlns.com/foaf/0.1/Person> ;
    <http://xmlns.com/foaf/0.1/firstName> "Sebastian" ;
    <http://xmlns.com/foaf/0.1/gender> "male" ;
    _:genid4 <http://www.w3.org/1999/02/22-rdf-syntax-ns#first> "shinobu" .
_:genid4 <http://www.w3.org/1999/02/22-rdf-syntax-ns#rest> <http://www.w3.org/1999/02/22-rdf-syntax-ns#nil> .
_:genid5 <http://www.w3.org/1999/02/22-rdf-syntax-ns#first> "Ruby" .
_:genid5 <http://www.w3.org/1999/02/22-rdf-syntax-ns#rest> _:genid4 .
_:genid6 <http://www.w3.org/1999/02/22-rdf-syntax-ns#first> "Jekyll-RDF" .
_:genid6 <http://www.w3.org/1999/02/22-rdf-syntax-ns#rest> _:genid5 .
_:genid7 <http://www.w3.org/1999/02/22-rdf-syntax-ns#first> "Informatik" .
_:genid7 <http://www.w3.org/1999/02/22-rdf-syntax-ns#rest> _:genid6 .
<http://example.org/knows> <http://www.w3.org/1999/02/22-rdf-syntax-ns#first> "BSV" .
<http://example.org/knows> <http://www.w3.org/1999/02/22-rdf-syntax-ns#rest> _:genid7 .
<http://example.org/Sebastian> <http://xmlns.com/foaf/0.1/knows> <http://example.org/knows> .
```

The template:

```
---
---
<h1 class="page-heading">{{ page.rdf | rdf_property: '<http://xmlns.com/foaf/0.1/name>' }}</h1>

<dl>
    {% assign firstname = page.rdf | rdf_property: '<http://xmlns.com/foaf/0.1/firstName>' %}
    {% assign lastname = page.rdf | rdf_property: '<http://xmlns.com/foaf/0.1/lastName>' %}
    {% assign gender = page.rdf | rdf_property: '<http://xmlns.com/foaf/0.1/gender>' %}
    {% if firstname %}
    <dt>Vorname:</dt>
    <dd>{{ firstname }}</dd>
    {% endif %}
    {% if lastname %}
    <dt>Nachname:</dt>
    <dd>{{ lastname }}</dd>
    {% endif %}
    {% if lastname %}
    <dt>Geschlecht:</dt>
    <dd>{{ gender }}</dd>
    {% endif %}
</dl>


<ul>
    {% assign coll = page.rdf | rdf_property: '<http://xmlns.com/foaf/0.1/knows>'  %}
    <span>{{ coll }}</span>
    {% assign array = coll | rdf_collection %}
    {% for know in array %}
    <li>{{ know }}</li>
    {% endfor %}
</ul>
```

Exception:

      Liquid Exception: undefined method `statements_as_subject' for nil:NilClass in rdfsites/http/xmlns.com/foaf/0.1/Agent/index.html

### PR DESCRIPTION
**Pull request by Simaris**
Fix #66, fix #10

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -1,3 +1,3 @@
 source 'https://rubygems.org'
-
+gem 'sparql', :git => 'https://github.com/ruby-rdf/sparql.git', :branch => 'develop'
```
Comment (white-gecko) [inline Gemfile]: This is not needed anymore the maintainer did a release: https://rubygems.org/gems/sparql/versions/2.2.1

**Comment by white-gecko** [comment]
The branch needs to be rebased.

**Comment by white-gecko** [comment]
This feature currently fails with:

```
n-triple:
 <http://example.org/knows> 
  Liquid Exception: undefined method `f' for #<RDF::Query::Solution:0x2b053aafa4b4({})> in Sebastian/index.html
jekyll 3.3.0 | Error:  undefined method `f' for #<RDF::Query::Solution:0x2b053aafa4b4({})>
```

with this graph:

```
@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .

<http://example.org/Sebastian>
    a <http://xmlns.com/foaf/0.1/Agent>, <http://xmlns.com/foaf/0.1/Person> ;
    <http://xmlns.com/foaf/0.1/firstName> "Sebastian" .

_:genid4 <http://www.w3.org/1999/02/22-rdf-syntax-ns#first> "somebody" .
_:genid4 <http://www.w3.org/1999/02/22-rdf-syntax-ns#rest> <http://www.w3.org/1999/02/22-rdf-syntax-ns#nil> .
_:genid5 <http://www.w3.org/1999/02/22-rdf-syntax-ns#first> "Ruby" .
_:genid5 <http://www.w3.org/1999/02/22-rdf-syntax-ns#rest> _:genid4 .
_:genid6 <http://www.w3.org/1999/02/22-rdf-syntax-ns#first> "Jekyll-RDF" .
_:genid6 <http://www.w3.org/1999/02/22-rdf-syntax-ns#rest> _:genid5 .
_:genid7 <http://www.w3.org/1999/02/22-rdf-syntax-ns#first> "Informatik" .
_:genid7 <http://www.w3.org/1999/02/22-rdf-syntax-ns#rest> _:genid6 .
<http://example.org/knows> <http://www.w3.org/1999/02/22-rdf-syntax-ns#first> "something" .
<http://example.org/knows> <http://www.w3.org/1999/02/22-rdf-syntax-ns#rest> _:genid7 .
<http://example.org/Sebastian> <http://xmlns.com/foaf/0.1/knows> <http://example.org/knows> .
```

and this `default_layout`:

```
---
---
<h1 class="page-heading">{{ page.rdf | rdf_property: '<http://xmlns.com/foaf/0.1/name>' }}</h1>

<dl>
    {% assign firstname = page.rdf | rdf_property: '<http://xmlns.com/foaf/0.1/firstName>' %}
    {% assign lastname = page.rdf | rdf_property: '<http://xmlns.com/foaf/0.1/lastName>' %}
    {% assign gender = page.rdf | rdf_property: '<http://xmlns.com/foaf/0.1/gender>' %}
    {% if firstname %}
    <dt>Vorname:</dt>
    <dd>{{ firstname }}</dd>
    {% endif %}
    {% if lastname %}
    <dt>Nachname:</dt>
    <dd>{{ lastname }}</dd>
    {% endif %}
    {% if lastname %}
    <dt>Geschlecht:</dt>
    <dd>{{ gender }}</dd>
    {% endif %}
</dl>

<ul>
    {% assign coll = page.rdf | rdf_property: '<http://xmlns.com/foaf/0.1/knows>'  %}
    {% if coll %}
    <span>{{ coll }}</span>
    {% assign array = coll | rdf_collection %}
    {% for know in array %}
    <li>{{ know }}</li>
    {% endfor %}
    {% endif %}
</ul>
```

File: lib/jekyll/filters/rdf_collection.rb [hunk 2]
```diff
@@ -26,30 +26,14 @@
...
-        return results
-      end
+      Jekyll.logger.info("n-triple:\n #{input.term.to_ntriples}")
+      sparql_query = input.sparql
+      query = "SELECT ?f WHERE{ #{input.term.to_ntriples} <http://www.w3.org/1999/02/22-rdf-syntax-ns#rest>* ?r. ?r <http://www.w3.org/1999/02/22-rdf-syntax-ns#first> ?f}"
+      results = []
+      sparql_query.query(query).each{ |solution|
+        results.push Jekyll::Drops::RdfTerm.build_term_drop(solution.f, input.sparql, input.site).addNecessities(input.site, input.page)
+      }
+      return results
```
Comment (white-gecko) [inline lib/jekyll/filters/rdf_collection.rb]: I think the issue is somewhere here.

**Comment by white-gecko** [comment]
Ok, the issue with `undefined method `f' for #<RDF::Query::Solution:0x2b053aafa4b4({})>` is solved with sparql lib 2.2.1

**Comment by white-gecko** [comment]
@Simaris looks good, will you also remove the dependency to the git repository and set sparql to 2.2.1 in the gemspec?

**Comment by Simaris** [comment]
Already done that, just pushed prematurely

**Comment by white-gecko** [comment]
@Simaris I've done some history rewriting to make the changes more compact. Further I've also made the sparql version constraint less restrictive. Is this ok

**Comment by Simaris** [comment]
It seems correct to me

**Pull request merged**